client/client_update.go

```diff
@@ -64,22 +66,44 @@ func (u *UpdateClient) GetScheduledUpdate(api ApiRequester, server string,
 	return u.getUpdateInfo(api, processUpdateResponse, server, current)
 }
 
+// getUpdateInfo Tries to get the next update information from the backend. This
+// is done in two stages. First it tries a POST request with the devices provide
+// parameters. Then if this fails with a 404 response, then it falls back to the
+// open source version with GET, and the parameters encoded in the URL.
 func (u *UpdateClient) getUpdateInfo(api ApiRequester, process RequestProcessingFunc,
 	server string, current CurrentUpdate) (interface{}, error) {
-	req, err := makeUpdateCheckRequest(server, current)
+	ent_req, req, err := makeUpdateCheckRequest(server, current)
 	if err != nil {
 		return nil, errors.Wrapf(err, "failed to create update check request")
 	}
 
-	r, err := api.Do(req)
-
+	r, err := api.Do(ent_req)
 	if err != nil {
-		log.Debug("Sending request error: ", err)
-		return nil, errors.Wrapf(err, "update check request failed")
+		log.Debugf("Failed sending device provides to the backend: Error: %v", err)
+		return nil, errors.Wrapf(err, "enterprise update check request failed")
 	}
 
 	defer r.Body.Close()
 
+	if r.StatusCode != http.StatusOK && r.StatusCode != http.StatusNoContent {
+		if r.StatusCode == http.StatusNotFound {
+			// 404 - Fall back to the old GET request
```

Comment:
@kjaskiewiczz: What do you think about this logic? The idea is that after the `deployments/next` POST request, it will fall back to a GET request if, and only if, the status code is 404.

Which was my suggestion. But now I'm a bit uneasy. Is it resilient enough? Can proxies mess this up, for example, by returning a different error code? What do you think?

Reply:
the logic itself looks ok, but I don't understand the idea of falling back to the GET request

Reply:
@kjaskiewiczz how would you want it done from the client side?

Reply:
> I don't understand the idea of falling back to the GET request

@kjaskiewiczz: The client doesn't know whether it's talking to an Enterprise or Open Source server. It has to fall back to the Open Source GET method to keep working with Open Source servers.

Reply:
right! thanks for clarification
proxies shouldn't modify 404 response

Reply:
Ok, thanks!